myhpom/models/document.py

```diff
@@ -75,8 +78,64 @@ def render_images(self, allpages=True, res=PDF_RESOLUTION, **gsargs):
         image_filenames = gs.render(pdf_filename, res=res, allpages=allpages, **gsargs)
         return image_filenames
 
-    def __str__(self):
-        return self.document.name
+    @property
+    def verification_in_progress(self):
+        """ Returns True when the CF for this AD is still in progress. """
+        # Search for any run that is in progress
+        return CloudFactoryDocumentRun.objects.filter(
+            status=CloudFactoryDocumentRun.STATUS_PROCESSING,
+            document_url__advancedirective=self).exists()
+
+    @property
+    def verification_failed(self):
+        """ Returns True when the CF task has failed.
+
+        A failure of the task (aborted, network problems, bad data at the
+        protocol level, etc) -- the verification process did not happen at this
+        stage (it failed before it could)
+        """
+        # Search for any run that is in progress
+        return CloudFactoryDocumentRun.objects \
+            .filter(
+                status__in=CloudFactoryDocumentRun.STATUS_FINAL_STATES,
+                document_url__advancedirective=self) \
+            .exclude(status=CloudFactoryDocumentRun.STATUS_PROCESSED) \
+            .exists()
+
+    @property
+    def verification_result(self):
+        """
+        Returns a dictionary of keys corresponding to the output of the
+        finished CF run.
+
+        If the CF verification is not finished or failed, returns None.
+        """
+        # Select all finished runs - for the first one that processed (there
+        # should be only one), return its output
+        processed_run = CloudFactoryDocumentRun.objects.filter(
+            status=CloudFactoryDocumentRun.STATUS_PROCESSED,
+            document_url__advancedirective=self).first()
```

Comment:
Good question - initially I'd implemented this as a 'look at all processed' runs for a document. But then I rethought it...shouldn't there only ever be one `STATUS_PROCESSED` or `STATUS_PROCESSING` run: if/when we reschedule a run b/c of a previous failure doesn't that mean the previous run(s) will be in a terminal state (DELETED, REQ_ERROR, etc)?

Reply:
Ah, yes, you're right – any other state will have been filtered out. I guess I'd still be more comfortable with .last(), but that's only because I don't trust the future.

Reply:
A scholarly thing to do - the future is never what we think! what about..not filtering and just doing 'get()' rather than first/last and changing the rest of these similar methods to use get() rather than exists()?

Then we'd have a fail-fast system - in the unlikely event that we have multiple of these records that should only be one - we get an exception, and an email?